### PR TITLE
fix(filesystem): UNC paths are not supported

### PR DESCRIPTION
**Pull request by IlyaFaer**
Towards #1175

### CODE REVIEW

**Comment by IlyaFaer** [comment]
@rudolfix, that's a couple of cunning fixes. I'd say I didn't find ideological errors - it's just about a couple of symbols (normal ones), which make `fsspec` (not our code) to turn the path into something weird. Can't say I like it, but it worked. Pushed a test in `verified-sources`: https://github.com/dlt-hub/verified-sources/pull/423

**Review by rudolfix — COMMENTED** [review]
@IlyaFaer it is hard to say if this works or not without comments and tests or maybe I'm reviewing too early?

**Review by rudolfix — CHANGES_REQUESTED** [review]
OK good!

but you can add tests also here. we are testing `glob` in `test_local_fileststem`. please unit test this
- UNC path
- windows path starting with drive letter
- windows path using backslashes

**Review by rudolfix — CHANGES_REQUESTED** [review]
more specific on the test, following https://en.wikipedia.org/wiki/File_URI_scheme
please test following windows paths

* C:\a\b\b
* \\localhost\c$\a
* file://server/folder/data.xml
* file:////server/folder/data.xml (if the above works, you can skip this one)
 with the `glob` function

**Comment by IlyaFaer** [comment]
@rudolfix, hm, I don't think it actually works. Surprisingly, when you try to read one file, it goes fine. But if you give a glob, `fsspec` does path translation of some sort, and always adds a disk name into the start, creating a pattern like:

`(?s:[\\\\/][\\\\/]localhost[\\\\/][\\\\/]C\\$[\\\\/]git_reps[\\\\/]dlt[\\\\/]tests[\\\\/]common[\\\\/]storages[\\\\/]samples[\\\\/].*)\\Z`

Here:
https://github.com/fsspec/filesystem_spec/blob/37c1bc63b9c5a5b2b9a0d5161e89b4233f888b29/fsspec/utils.py#L742

So, basically, it can see files by UNC path (I logged, it really does), but it filters them out, not returning anything. Not sure if it can be overridden...

However, if I give a file path (not a glob), it reads the file correctly, not going to this part of code, which does the path translation.

If I give a normal full path, like `C:/a/b/c`, it works fine. But UNC - no.

**Comment by IlyaFaer** [comment]
@rudolfix, the only way I see how to fix it, is to write our own filesystem, inheriting the `fsspec.LocalFileSystem` and override the `glob` method to fix this translation. As there are no conditions in the translation, there is no way to skip it for globs (it's skipped only for precise paths). What do you think?

**Comment by rudolfix** [comment]
@IlyaFaer 

> @rudolfix, the only way I see how to fix it, is to write our own filesystem, inheriting the fsspec.LocalFileSystem and override the glob method to fix this translation. As there are no conditions in the translation, there is no way to skip it for globs (it's skipped only for precise paths). What do you think?

heh this is so bad :) could you check first if regular Python glob works correctly? then my take would be to use it for `file://` and just translate the results to something fsspec understand. but let's check it out.

**Comment by IlyaFaer** [comment]
@rudolfix, yes, Python works fine, I tried it even before started to look at `fsspec`

Untitled

**Comment by IlyaFaer** [comment]
> translate the results to something fsspec understand

I don't think it's possible. Because, if we read `localhost`, it's okay, we can translate the path. But what if we read a network entity? It'll start with an ip address - it can't be translated into a path, which starts with a disk name. So, `fsspec` will add it into the pattern by itself.

File: tests/load/filesystem/test_filesystem_common.py [hunk 1]
```diff
@@ -256,3 +258,42 @@ def test_filesystem_destination_passed_parameters_override_config_values() -> No
...
+                "freshman_lbs.csv",
+                "mlb_players.csv",
+                "mlb_teams_2012.csv",
+                "mlb_players.jsonl",
+                "A881_20230920.csv",
+                "A803_20230919.csv",
+                "A803_20230920.csv",
+                "mlb_players.parquet",
+                "taxi.csv.gz",
+                "sample.txt",
```
Comment (IlyaFaer) [inline tests/load/filesystem/test_filesystem_common.py:293]: When working with UNC, paths are always full and absolute (to include the IP address of the host), so it requires a bit different test case.

File: dlt/common/storages/fsspec_filesystem.py [hunk 2]
```diff
@@ -282,13 +283,22 @@ def glob_files(
         # this is a file so create a proper file url
         bucket_url = pathlib.Path(bucket_url).absolute().as_uri()
         bucket_url_parsed = urlparse(bucket_url)
+
     bucket_url_no_schema = bucket_url_parsed._replace(scheme="", query="").geturl()
     bucket_url_no_schema = (
         bucket_url_no_schema[2:] if bucket_url_no_schema.startswith("//") else bucket_url_no_schema
     )
     filter_url = posixpath.join(bucket_url_no_schema, file_glob)
+    is_unc_path = fs_client.protocol == "file" and "$" in filter_url
```
Comment (rudolfix) [inline dlt/common/storages/fsspec_filesystem.py:292]: I started to experiment with this place but it still does not work. all paths (bucket_url) and filter_url are normalized and $ is never there - at least for your test cases

File: dlt/common/storages/fsspec_filesystem.py [hunk 3]
```diff
@@ -302,10 +312,16 @@ def glob_files(
         # make that absolute path on a file://
         if bucket_url_parsed.scheme == "file" and not file.startswith("/"):
             file = f"/{file}"
-        file_name = posixpath.relpath(file, bucket_url_no_schema)
-        file_url = bucket_url_parsed._replace(
-            path=posixpath.join(bucket_url_parsed.path, file_name)
-        ).geturl()
+
+        if is_unc_path:
+            file_name = file
```
Comment (rudolfix) [inline dlt/common/storages/fsspec_filesystem.py:317]: name must contain only name (final path component)

File: tests/common/storages/test_local_filesystem.py [hunk 4]
```diff
@@ -54,3 +55,41 @@ def test_filesystem_decompress() -> None:
...
         with file_dict.open(compression="enable") as f:
             assert f.read().startswith(b'"1200864931","2015-07-01 00:00:13"')
+
+
+@pytest.mark.skipif(platform.system() != "Windows", reason="Test it only on Windows")
+def test_windows_unc_path() -> None:
+    config = FilesystemConfiguration(bucket_url=TEST_SAMPLE_FILES)
+    config.read_only = True
+
+    unc_path = r"\\localhost\\" + os.path.abspath("tests/common/storages/samples").replace(":", "$")
```
Comment (rudolfix) [inline tests/common/storages/test_local_filesystem.py:65]: this returns no elements in the glob

File: tests/common/storages/test_local_filesystem.py [hunk 5]
```diff
@@ -54,3 +55,41 @@ def test_filesystem_decompress() -> None:
...
+                "mlb_teams_2012.csv",
+                "mlb_players.jsonl",
+                "A881_20230920.csv",
+                "A803_20230919.csv",
+                "A803_20230920.csv",
+                "mlb_players.parquet",
+                "taxi.csv.gz",
+                "sample.txt",
+            ]
+            for file in all_file_items:
```
Comment (rudolfix) [inline tests/common/storages/test_local_filesystem.py:91]: if there are no elements this tests still passes

File: tests/common/storages/test_local_filesystem.py [hunk 6]
```diff
@@ -54,3 +55,41 @@ def test_filesystem_decompress() -> None:
...
+                file_name = file["file_name"].split("\\")[-1].split("/")[-1]
+                assert file_name in expected_files
+        except NotImplementedError as ex:
+            pytest.skip(f"Skipping due to {str(ex)}")
```
Comment (rudolfix) [inline tests/common/storages/test_local_filesystem.py:95]: what does it mean?

**Review by rudolfix — CHANGES_REQUESTED** [review]
there are things that still does not work

File: tests/common/storages/test_local_filesystem.py (hunk 5 above)
Comment (IlyaFaer) [reply tests/common/storages/test_local_filesystem.py:91]: Oh, yeah, it is. There were a couple of small changes, which I reverted, because tests were till passing without them 😵‍💫Fixed

File: dlt/common/storages/fsspec_filesystem.py [hunk 7]
```diff
@@ -276,19 +279,30 @@ def glob_files(
     """
     import os
 
+    is_unc_path = "$" in bucket_url
     bucket_url_parsed = urlparse(bucket_url)
     # if this is a file path without a scheme
-    if not bucket_url_parsed.scheme or (os.path.isabs(bucket_url) and "\\" in bucket_url):
+    if (
+        not bucket_url_parsed.scheme or (os.path.isabs(bucket_url) and "\\" in bucket_url)
+    ) and not is_unc_path:
```
Comment (IlyaFaer) [inline dlt/common/storages/fsspec_filesystem.py:287]: This part must not be done for UNC, as it breaks them completely

File: dlt/common/storages/fsspec_filesystem.py [hunk 8]
```diff
@@ -299,20 +313,28 @@ def glob_files(
...
         if bucket_url_parsed.scheme == "file" and not file.startswith("/"):
             file = f"/{file}"
-        file_name = posixpath.relpath(file, bucket_url_no_schema)
-        file_url = bucket_url_parsed._replace(
-            path=posixpath.join(bucket_url_parsed.path, file_name)
-        ).geturl()
+
+        if is_unc_path:
+            file_url = "file:///" + file
+            file_name = file.replace(bucket_url_no_schema, "").replace("\\", "/").lstrip("/")
```
Comment (IlyaFaer) [inline dlt/common/storages/fsspec_filesystem.py:323]: It's the most straightforward and stupid solution, but I think it'll work. For UNC we're getting `file` from `glob`, which is supposed to always return the same result, so `str` edits like this should work. This way it produces the same result as `relpath` for normal paths.

File: dlt/common/storages/fsspec_filesystem.py (hunk 3 above)
Comment (rudolfix) [reply dlt/common/storages/fsspec_filesystem.py]: good point! although our docs mention that file_name can be a relative path, it is very misleading. so let's fix it:
1. rename `file_name` to `relative_path` and add `file_name` with just the name part
2. update all the tests
3. update documentation: https://dlthub.com/docs/dlt-ecosystem/verified-sources/filesystem#fileitem-representation

luckily we do not use `file_name` in verified sources

File: tests/common/storages/test_local_filesystem.py [hunk 9]
```diff
@@ -54,3 +55,39 @@ def test_filesystem_decompress() -> None:
...
+            "met_csv/A803/A803_20230919.csv",
+            "met_csv/A803/A803_20230920.csv",
+            "parquet/mlb_players.parquet",
+            "gzip/taxi.csv.gz",
+            "sample.txt",
+        ]
+        assert len(all_file_items) == len(expected_files)
+
+        for file in all_file_items:
+            assert file["file_name"] in expected_files
```
[the remark this answers is not in the document]
Comment (rudolfix) [reply tests/common/storages/test_local_filesystem.py:93]: this will also try to load the content of the file so it will verify if fsspec `open` works. I think it is crucial here

File: dlt/common/storages/fsspec_filesystem.py (hunk 8 above)
Comment (rudolfix) [reply dlt/common/storages/fsspec_filesystem.py:323]: what is `file` here? file_url must contain UNC path so we are able to use it to open later. make sure it is like that

**Review by rudolfix — CHANGES_REQUESTED** [review]

File: tests/common/storages/test_local_filesystem.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (IlyaFaer) [reply tests/common/storages/test_local_filesystem.py]: @rudolfix, opening doesn't work. Hm-m, I didn't notice it, but our code uses `fsspec.open()`, which (just like any other `fsspec` method) break UNC paths. O-o-okay, I added like a separate branch for UNC. Locally, the tests all passed, pushing (there are still several small things I didn't do, but I wanna see if I didn't break anything).

**Comment by IlyaFaer** [comment]
@rudolfix, uf-f, that's quite a big re-writing. But I think I understood it

**Review by rudolfix — APPROVED** [review]

**Comment by rudolfix** [comment]
@IlyaFaer it was actually impossible to correctly implement all file paths styles on all platforms using fsspec. what this PR does is to use Python `glob` for all `file://` uris. it also converts all `file://` uris to native paths before opening. somehow fsspec is able to open UNC paths, but it is not able to do that when they are in `file://` form.

as usual tests are way more complicated than code...

File: dlt/common/storages/configuration.py [hunk 10]
```diff
@@ -117,11 +116,57 @@ def fingerprint(self) -> str:
...
+                # filesystem root
+                if local_path == "/":
+                    return str(pathlib.Path("/").resolve())
+                # this prevents /C:/ or ///share/ where both POSIX and Windows root are present
+                if os.path.isabs(local_path[1:]):
+                    local_path = local_path[1:]
+        return str(pathlib.Path(local_path))
+
+    @staticmethod
+    def make_file_uri(local_path: str) -> str:
```
[the remark this answers is not in the document]
Comment (rudolfix) [reply dlt/common/storages/configuration.py:164]: you are right that this needs a better place. we have file utils in `FileStorage` but they probably need to go somewhere else, then we'll move those as well
